src/equations/equations.jl

```diff
@@ -176,6 +176,26 @@ struct BoundaryConditionWall{B}
   boundary_value_function::B
 end
 
+# Wall boundary condition for use with TreeMesh or StructuredMesh
+@inline function (boundary_condition::BoundaryConditionWall)(u_inner, orientation_or_normal,
+                                                             direction,
+                                                             x, t,
+                                                             surface_flux_function, equations)
+
+  u_boundary = boundary_condition.boundary_value_function(u_inner, orientation_or_normal, equations)
+
+  # Calculate boundary flux
+  if direction in (2, 4, 6) # u_inner is "left" of boundary, u_boundary is "right" of boundary
```

Comment:
```suggestion
  if iseven(direction) # u_inner is "left" of boundary, u_boundary is "right" of boundary
```
@ranocha I've seen you changing that in some other PR. Was that only for aesthetic reasons, or did it improve the performance for some reason?

Reply:
Good question. At least to me, it's a definite improvement in terms of comprehensibility: The original version might make someone wonder what's special about `2`, `4`, and `6`, while the suggestion by @ranocha self-descriptively states that we are only interested in the "even number" property and not the specific values.

I did a performance test, because I was curious, and it turns out `iseven` can be significantly faster:
<details>
<summary>performance measurements</summary>

```julia
julia> using BenchmarkTools

julia> f(direction) = in(direction, (2, 4, 6))
f (generic function with 1 method)

julia> g(direction) = iseven(direction)
g (generic function with 1 method)

julia> d = 3
3

julia> @benchmark f($d)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  3.281 ns … 18.712 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.274 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.339 ns ±  0.576 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                           
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁
  5.38 ns        Histogram: frequency by time        3.81 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark g($d)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  0.023 ns … 4.491 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     0.026 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   0.028 ns ± 0.045 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                          
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁
  0.051 ns       Histogram: frequency by time      0.025 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> d = 2
2

julia> @benchmark f($d)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  0.022 ns … 0.056 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     0.026 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   0.027 ns ± 0.003 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                          
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁
  0.04 ns        Histogram: frequency by time      0.032 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark g($d)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  0.022 ns … 0.085 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     0.025 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   0.026 ns ± 0.002 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                          
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁ ▁
  0.048 ns       Histogram: frequency by time      0.025 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> d = 6
6

julia> @benchmark f($d)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.881 ns … 17.867 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     3.034 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   3.145 ns ±  0.414 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                           
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁ ▁
  3.74 ns        Histogram: frequency by time        2.89 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark g($d)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  0.022 ns … 0.118 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     0.026 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   0.026 ns ± 0.003 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                          
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁ ▁
  0.058 ns       Histogram: frequency by time      0.025 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
</details>

Gist: While `iseven` has near-constant timings for any integer, the speed of `in` is value dependent. Unless I've made a mistake setting up the performance measurements and allowed Julia to cheat for some of the tests :thinking:

Reply:
It's clearly expected to be faster since you only need to check the sign bit instead of looping through all values in the tuple

Reply:
@ranocha = :man_teacher:
@sloede = :open_mouth: --> :thinking: --> :bulb: --> :facepalm:

Reply:
> Unless I've made a mistake setting up the performance measurements and allowed Julia to cheat for some of the tests

Well, you did - every time you see sub-nanosecond benchmarks, you can be sure that you're not measuring any computation (unless you overclocked your CPU by a factor of 1000). You should wrap the values in `Ref` if you want to benchmark a single value, i.e.,
```julia
julia> @benchmark f($(Ref(3))[])
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.851 ns … 20.414 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.931 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.966 ns ±  0.675 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                           
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁ ▁
  3.08 ns        Histogram: frequency by time        2.93 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark g($(Ref(3))[])
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.058 ns … 16.046 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.060 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.063 ns ±  0.154 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                           
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁  ▁
  1.21 ns        Histogram: frequency by time        1.06 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
Alternatively, you can also benchmark random values using something like 
```julia
julia> @benchmark f(direction) setup=(direction=rand((1,2,3,4,5,6)))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.266 ns … 17.620 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.725 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.161 ns ±  0.818 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                        ▁ ▂
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ █
  2.84 ns      Histogram: log(frequency) by time     2.93 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark g(direction) setup=(direction=rand((1,2,3,4,5,6)))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.057 ns … 16.033 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.061 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.090 ns ±  0.606 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                           
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁  ▁
  1.14 ns        Histogram: frequency by time        1.06 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Reply:
Thanks a lot for this lesson, this is really helpful! :clap: